service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/UpdateAiNotebookAttributesStep.java

```
@@ -0,0 +1,108 @@
+
+  UpdateAiNotebookAttributesStep(
+      ControlledAiNotebookInstanceResource resource,
+      CrlService crlService,
+      GcpCloudContextService gcpCloudContextService
+  ) {
+    this.resource = resource;
+    this.crlService = crlService;
+    this.cloudContextService = gcpCloudContextService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final FlightMap inputMap = context.getInputParameters();
+    final ApiGcpAiNotebookUpdateParameters updateParameters =
+        inputMap.get(UPDATE_PARAMETERS, ApiGcpAiNotebookUpdateParameters.class);
+
+    return updateAiNotebook(updateParameters);
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    final FlightMap workingMap = context.getWorkingMap();
+    final ApiGcpAiNotebookUpdateParameters prevParameters =
+        workingMap.get(PREVIOUS_UPDATE_PARAMETERS, ApiGcpAiNotebookUpdateParameters.class);
+    var projectId =
+        cloudContextService.getRequiredGcpProject(resource.getWorkspaceId());
+    InstanceName instanceName = resource.toInstanceName(projectId);
+    AIPlatformNotebooksCow notebooks = crlService.getAIPlatformNotebooksCow();
+    try {
+      var currentMetadata = notebooks.instances().get(instanceName).execute().getMetadata();
+      for (var entry: currentMetadata.entrySet()) {
+        if (!prevParameters.getMetadata().containsKey(entry.getKey())) {
+          // reset the new key entry to "" value because the gcp api does not allow deleting
+          // metadata item so we can't simply undo the add.
+          currentMetadata.put(entry.getKey(), "");
+        }
+      }
+      return updateAiNotebook(new ApiGcpAiNotebookUpdateParameters().metadata(currentMetadata));
+    } catch (GoogleJsonResponseException e) {
+      if (HttpStatus.BAD_REQUEST.value() == e.getStatusCode() || HttpStatus.NOT_FOUND.value() == e.getStatusCode()) {
```

I think this should throw an exception and fail the flight to be consistent with notebook creation. 

currently when creating the notebook, we throw a `ReservedMetadataKeyException` and fail the operation if the user specifies a reserved key. In this case, we're silently not including it and continuing. I think either approach is reasonable, but we should at least be consistent.

---

i'm just forseeing that it gives more burden on the UI  to check for valid input :) but yeah i agree consistency is good.